GiniVision/Classes/Core/CameraViewController.swift

```diff
@@ -614,15 +651,56 @@ extension CameraViewController {
     }
 }
 
+// MARK: - DocumentPickerCoordinatorDelegate
+
+extension CameraViewController: DocumentPickerCoordinatorDelegate {
+
+    func documentPicker(_ coordinator: DocumentPickerCoordinator,
+                        didPick documents: [GiniVisionDocument],
+                        from: DocumentPickerType,
+                        completion: DocumentPickerCompletion?) {
+        let loadingView = addValidationLoadingView()
+        self.validate(importedDocuments: documents) { validatedDocuments in
+            loadingView.removeFromSuperview()
+            let elementsWithError = validatedDocuments.filter { $0.1 != nil}
+            if let firstElement = elementsWithError.first,
+                let error = firstElement.1,
+                (!self.giniConfiguration.multipageEnabled || firstElement.0.type != .image) {
+                if let completion = completion {
+                    completion(nil) {
+                        self.showErrorDialog(for: error)
+                    }
+                } else {
+                    self.showErrorDialog(for: error)
+                }
+            } else {
+                self.process(validatedImportedDocuments: validatedDocuments.map {$0.0 }) { [weak self] error, didDismiss in
+                    guard let `self` = self else { return }
+                    // This is needed since the `UIDocumentPickerViewController` is automatically
+                    // dismissed and the drag&drop is done in this view controller.
+                    if from != .gallery {
+                        guard let error = error else {
+                            didDismiss?()
+                            return
+                        }
+                        
+                        self.showErrorDialog(for: error)
+                    } else {
+                        completion?(error, didDismiss)
```

Comment:
Look at this (the else statement) and the if statement proceeding it. There are so many cases. That's why I complained that you have a completion handler within a completion handler. It's a rabbit hole. And it's endless.

Reply:
I simplified it a little bit and also renamed the closure to make them more clear.